### PR TITLE
PCHR-3251: Upgrader for single day requests in hours

### PR DESCRIPTION
**Pull request by igorpavlov**
## Overview
As per https://github.com/civicrm/civihr/pull/2443 end time can be chosen for single day leave requests in hours calculation unit. However, existing such leave requests are storing end time currently equal to start time. This upgrader fetches such leave requests, retrieves information about work pattern day basing on the date of the leave request and, if it is still a working day, sets the end time from the working day as an end time of the leave request.

## Before/After

upg

## Technical Details

The upgrader simply takes "to" date of the single day request (which equals the "from" date) and  replaces the time taken from the work pattern day corresponding to this date.

```php
// Construct query that selects all existing single day leave requests in hours
$query = "SELECT leaveRequest.id
  FROM {$leaveRequestTable} leaveRequest
  INNER JOIN {$absenceTypeTable} absenceType ON leaveRequest.type_id = absenceType.id
  WHERE
    absenceType.calculation_unit = {$calculationUnits['hours']}
      AND leaveRequest.from_date IS NOT NULL
      AND leaveRequest.from_date = leaveRequest.to_date
      AND leaveRequest.is_deleted = 0";

// ...

while ($leaveRequestRecord->fetch()) {
  // Get a leave request instance by ID
  $leaveRequest = LeaveRequest::findById($leaveRequestRecord->id);
  // Get a work day info based on the contact of the leave request and "from" date
  $workDay = ContactWorkPattern::getContactWorkDayForDate(
    $leaveRequest->contact_id, new DateTime($leaveRequest->from_date));

  // If this is still a working day and end time exists
  if (!empty($workDay['time_to'])) {
    // Simply replace the request end time with the one from the work pattern day
    $leaveRequest->to_date =
      substr($leaveRequest->from_date, 0, 10) . ' ' . $workDay['time_to'] . ':00';
    // And save
    $leaveRequest->save();
  }
};
```

### Upgrader performance

The upgrader was tested on a database of sample requests: 4000 requests to update and 10000 other requests (14K in total) and was executed in **16 sec** on a virtual machine with 4GB RAM (host machine 3.1 GHz Intel Core i7).

### Calendar tooltips

Additionally, calendar tooltip was fixed: it now shows "to" time as well as "from" time for single day requests in hours.

#### Before

image

#### After

image

### PHP unit tests

PHP unit tests in L&A passed both before and after upgrade.

---- 

✅ Manual Tests - passed
✅ PHP Unit Tests - passed
⏹ Backstop tests - omitted
⏹ CSS distributive files - not needed
⏹ Jasmine Tests - not needed
⏹ JS distributive files - not needed

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        absenceType.calculation_unit = {$calculationUnits['hours']}
+          AND leaveRequest.from_date IS NOT NULL
+          AND leaveRequest.from_date = leaveRequest.to_date
+          AND leaveRequest.is_deleted = 0";
+
+    $leaveRequestRecord = CRM_Core_DAO::executeQuery($query);
+
+    while ($leaveRequestRecord->fetch()) {
+      $leaveRequest = LeaveRequest::findById($leaveRequestRecord->id);
+      $workDay = ContactWorkPattern::getContactWorkDayForDate(
```
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php]: This method is not a static method, An instance of the service class is needed first to be able to use the method.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    $calculationUnits = array_flip(AbsenceType::buildOptions('calculation_unit', 'validate'));
+
+    $leaveRequestTable = LeaveRequest::getTableName();
+    $absenceTypeTable = AbsenceType::getTableName();
+
+    $query = "SELECT leaveRequest.id
+      FROM {$leaveRequestTable} leaveRequest
+      INNER JOIN {$absenceTypeTable} absenceType ON leaveRequest.type_id = absenceType.id
+      WHERE
+        absenceType.calculation_unit = {$calculationUnits['hours']}
```
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php]: Can you move this line up with the WHERE statement and let the other lines align.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+use CRM_HRLeaveAndAbsences_BAO_LeaveRequest as LeaveRequest;
+use CRM_HRLeaveAndAbsences_Service_ContactWorkPattern as ContactWorkPattern;
+
+trait CRM_HRLeaveAndAbsences_Upgrader_Step_1018 {
+
+  /**
+   * As per PCHR-3164 end time can be chosen for single day leave requests
+   * in hours calculation unit. However, existing such leave requests are
+   * storing end time currently equal to start time. This upgrader fetches such
+   * leave requests, retrieves information about work pattern day basing on the
```
[the remark this answers is not in the document]
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php (hunk 2 above)
Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php]: Done.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php [hunk 4]
```diff
@@ -26,17 +26,17 @@ public function upgrade_1018 () {
     $query = "SELECT leaveRequest.id
       FROM {$leaveRequestTable} leaveRequest
       INNER JOIN {$absenceTypeTable} absenceType ON leaveRequest.type_id = absenceType.id
-      WHERE
-        absenceType.calculation_unit = {$calculationUnits['hours']}
-          AND leaveRequest.from_date IS NOT NULL
-          AND leaveRequest.from_date = leaveRequest.to_date
-          AND leaveRequest.is_deleted = 0";
+      WHERE absenceType.calculation_unit = {$calculationUnits['hours']}
+        AND leaveRequest.from_date IS NOT NULL
```
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php]: Please align the AND queries with the other sql statements

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php [hunk 5]
```diff
@@ -26,17 +26,17 @@ public function upgrade_1018 () {
...
+        AND leaveRequest.from_date = leaveRequest.to_date
+        AND leaveRequest.is_deleted = 0";
 
     $leaveRequestRecord = CRM_Core_DAO::executeQuery($query);
 
     while ($leaveRequestRecord->fetch()) {
       $leaveRequest = LeaveRequest::findById($leaveRequestRecord->id);
-      $workDay = ContactWorkPattern::getContactWorkDayForDate(
+      $contactWorkPatternService = new ContactWorkPatternService();
```
Comment (tunbola) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php]: Please instantiate the service outside the while loop so that we won't have to be creating an object each time the loop runs.

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+      WHERE absenceType.calculation_unit = {$calculationUnits['hours']}
+      AND leaveRequest.from_date IS NOT NULL
+      AND leaveRequest.from_date = leaveRequest.to_date
+      AND leaveRequest.is_deleted = 0";
+
+    $leaveRequestRecord = CRM_Core_DAO::executeQuery($query);
+    $contactWorkPatternService = new ContactWorkPatternService();
+
+    while ($leaveRequestRecord->fetch()) {
+      $leaveRequest = LeaveRequest::findById($leaveRequestRecord->id);
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php]: @igorpavlov since you're worried about performance, you can avoid this `findById()` call. Here's how you can do it:

```php
$query = "SELECT id FROM absence_type WHERE calculation_unit = {$calculationUnits['hours']}";
    $result = CRM_Core_DAO::executeQuery($query)->fetchAll();
    $absenceTypesInHour = array_column($result, 'id');

    $leaveRequest = new LeaveRequest();
    $leaveRequest->whereAdd('from_date IS NOT NULL');
    $leaveRequest->whereAdd('from_date = to_date');
    $leaveRequest->whereAdd('is_deleted = 0');
    $leaveRequest->whereAdd('type_id IN (' . implode(',', $absenceTypesInHour) . ')';
    $leaveRequest->find();

    $contactWorkPatternService = new ContactWorkPatternService();
    
    while ($leaveRequest->fetch()) {
      $workDay = $contactWorkPatternService->getContactWorkDayForDate(
        $leaveRequest->contact_id, new DateTime($leaveRequest->from_date));

      if (!empty($workDay['time_to'])) {
        $leaveRequest->to_date =
          substr($leaveRequest->from_date, 0, 10) . ' ' . $workDay['time_to'] . ':00';

        $leaveRequest->save();
      }
    }
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+      $workDay = $contactWorkPatternService->getContactWorkDayForDate(
+        $leaveRequest->contact_id, new DateTime($leaveRequest->from_date));
+
+      if (!empty($workDay['time_to'])) {
+        $leaveRequest->to_date =
+          substr($leaveRequest->from_date, 0, 10) . ' ' . $workDay['time_to'] . ':00';
+
+        $leaveRequest->save();
+      }
+    };
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php]: We don't need the `;` here

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+    $contactWorkPatternService = new ContactWorkPatternService();
+
+    while ($leaveRequestRecord->fetch()) {
+      $leaveRequest = LeaveRequest::findById($leaveRequestRecord->id);
+      $workDay = $contactWorkPatternService->getContactWorkDayForDate(
+        $leaveRequest->contact_id, new DateTime($leaveRequest->from_date));
+
+      if (!empty($workDay['time_to'])) {
+        $leaveRequest->to_date =
+          substr($leaveRequest->from_date, 0, 10) . ' ' . $workDay['time_to'] . ':00';
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php]: a slightly better version:

```php
date("Y-m-d {$workDay['time_to']}:00", strtotime($leaveRequest->from_date));
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php [hunk 9]
```diff
@@ -23,29 +23,31 @@ public function upgrade_1018 () {
     $leaveRequestTable = LeaveRequest::getTableName();
     $absenceTypeTable = AbsenceType::getTableName();
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php]: Now `$leaveRequestTable` is not being used anymore and can be removed

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php [hunk 10]
```diff
@@ -23,29 +23,31 @@ public function upgrade_1018 () {
...
+
+    $result = CRM_Core_DAO::executeQuery($query)->fetchAll();
+    $absenceTypesInHour = array_column($result, 'id');
+
+    $leaveRequest = new LeaveRequest();
+    $leaveRequest->whereAdd('from_date IS NOT NULL');
+    $leaveRequest->whereAdd('from_date = to_date');
+    $leaveRequest->whereAdd('is_deleted = 0');
+    $leaveRequest->whereAdd('type_id IN (' . implode(',', $absenceTypesInHour) . ')');
+    $number_of_rows = $leaveRequest->find();
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php]: You're not using this `$number_of_rows` variable anywhere. Please remove it

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php [hunk 11]
```diff
@@ -23,29 +23,31 @@ public function upgrade_1018 () {
...
       $workDay = $contactWorkPatternService->getContactWorkDayForDate(
         $leaveRequest->contact_id, new DateTime($leaveRequest->from_date));
 
       if (!empty($workDay['time_to'])) {
         $leaveRequest->to_date =
-          substr($leaveRequest->from_date, 0, 10) . ' ' . $workDay['time_to'] . ':00';
+          date("Y-m-d {$workDay['time_to']}:00", strtotime($leaveRequest->from_date));
 
-        $leaveRequest->save();
+        $leaveRequest->update();
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php]: Why the change from `save` to `update`? `save` will call the `update()` method if the object has an ID, which I believe is the case here

Comment (igorpavlov) [reply uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php]: Save() will not work here due to calling of ->free().

File: uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php [hunk 12]
```diff
@@ -27,11 +24,30 @@ public function upgrade_1018 () {
...
+   * Fetches single day leave requests in hours,
+   * retrieves information about work pattern day based on
+   * the date of the leave request and the contact and,
+   * if it is still a working day, sets the end time from
+   * the working day as an end time of the leave request.
+   *
+   * @param array $absenceTypesInHourIDs - comma separated list of IDs of
+   *   Absence Types with "hours" calculation unit
+   */
+  private function update_single_day_requests_in_hours_end_time ($absenceTypesInHourIDs) {
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrleaveandabsences/CRM/HRLeaveAndAbsences/Upgrader/Step/1018.php]: For private methods in Upgraders we use this pattern: `up<upgrade number>_nameInCamelCase()` (check [here](https://github.com/civicrm/civihr/blob/528028facc466128e44cf84116926b9d14cadfc8/uk.co.compucorp.civicrm.hrcore/CRM/HRCore/Upgrader/Steps/1001.php#L71) for an example). The reason is to avoid collisions with other traits or private methods in the classes using the trait.

If you don't mind having two returns in this method another possible implementation would be:

```php
if (empty($absenceTypesInHours)) {
   return true;
}

... do the update

return true;
```
